Feature/005-OPEN-API-FOOD

seokhyeong9171 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**

**TO-BE**
- 식단 기능을 담당할 meal-service 모듈을 생성하였습니다.
- 공공데이터에서 식품 정보를 가져오는 기능을 개발하였습니다.
    - open feign을 이용해 가져오도록 개발하였습니다.
    - Gson 라이브러리를 사용하여 json을 파싱하였습니다.
    - FoodEntity에 코드 값을 위한 필드를 추가하였습니다.

**피드백 적용**
- 공공데이터에서 가져오는 Json을 매핑할 클래스를 만들어 매핑된 결과를 가져오도록 변경하였습니다.
- 기존에는 데이터의 total num을 구하고 그에 맞춰 데이터를 가져왔지만 데이터의 갯수가 0개가 되면 데이터 가져오기를 중지하는 방식으로 바꾸었습니다.

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [ ] 테스트 코드
- [ ] API 테스트

seokhyeong9171 committed:
feature: meal-service module 생성

seokhyeong9171 committed:
feature: food entity, domainDto 수정 -> food code 추가

seokhyeong9171 committed:
feature: 공공데이터에서 식품 정보 가져오는 기능 개발

seokhyeong9171 committed:
refactor: json 가져오지 못할때 error 추가

seokhyeong9171 committed:
refactor: ERD 이미지 수정

seokhyeong9171 committed:
refactor: 주석 추가

mingming-mentor commented on meal-service/src/main/java/com/health/mealservice/service/FoodPublicDataComponent.java:
```diff
@@ -0,0 +1,140 @@
+  public Integer getTotalNum() {
+
+    String json =
+        foodPublicDataClient.getFoodList(serviceKey, 1, 1, "json").getBody();
```
1, 1은 어떤 값인가요?

seokhyeong9171 replied:
@mingming-mentor  총 데이터 수를 가져오기 위해 가장 적은 데이터량만 요청해서 json 값을 가져오기 위함이었습니다
피드백 주신대로 수정해보겠습니다!

mingming-mentor commented on meal-service/src/main/java/com/health/mealservice/client/FoodPublicDataClient.java:
```diff
@@ -0,0 +1,20 @@
+public interface FoodPublicDataClient {
+
+  @GetMapping("")
+  ResponseEntity<String> getFoodList(
```
String으로 받는 것보단 해당 API에서 제공하는 json 형태를 class로 만들어서 해당 class로 받아오시면 좋을 것 같아요. 
parsing을 깔끔하게 할 수 있고 휴먼 에러가 발생할 확률이 줄어들 것 같습니다.

mingming-mentor commented on meal-service/src/main/java/com/health/mealservice/service/FoodPublicDataService.java:
```diff
@@ -0,0 +1,36 @@
+  private final FoodPublicDataComponent foodPublicDataComponent;
+  private final FoodRepository foodRepository;
+
+  public void getFoodData() {
```
이 경우 getAndSaveFoodData 가 좀더 맞을 것 같아요. getXXX의 경우 return 값으로 데이터를 반환한다는 의미가 있고, save 로직까지 동작을 한다면 메소드명에 save까지 명시를 하는 것이 협업시에 불필요한 사이드 이펙트를 줄일 수 있을 듯 합니다

mingming-mentor commented on meal-service/src/main/java/com/health/mealservice/service/FoodPublicDataService.java:
```diff
@@ -0,0 +1,36 @@
+
+    for (int i = 1; i < lastNum + 1; i++) {
+
+      List<FoodPublicDataDto> foodInfoList = foodPublicDataComponent.getApiDtoList(i);
+
+      List<FoodEntity> entityList = foodInfoList.stream().map(FoodPublicDataDto::toEntity).toList();
+
+      foodRepository.saveAll(entityList);
```
`foodRepository.saveAll(foodPublicDataComponent.getApiDtoList(i).stream().map(FoodPublicDataDto::toEntity).toList());`
요렇게 하면 좀더 가독성이 좋을 것 같고, 불필요한 변수 선언이 줄어들 것 같아요

mingming-mentor commented on meal-service/src/main/java/com/health/mealservice/service/FoodPublicDataComponent.java:
```diff
@@ -0,0 +1,140 @@
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class FoodPublicDataComponent {
```
이 부분은 FoodPublicDataClient에 드린 코멘트를 반영하시면 코드가 훨씬 가독성이 높아질 것 같네요

mingming-mentor commented on meal-service/src/main/java/com/health/mealservice/service/FoodPublicDataService.java:
```diff
@@ -0,0 +1,36 @@
+  public void getFoodData() {
+
+    // 공공데이터에서 한번에 식품 정보 최대 100개 까지 가져올 수 있음
+    int totalNum = foodPublicDataComponent.getTotalNum();
```
total Num을 하나 보내기보단, 우선 요청을 하나 보내면서 아래 saveAll을 처리하고, 이때 응답으로 받은 totalNum을 이용하여 반복문을 동작하도록 하면 좋을 것 같아요. 
totalNum()을 보니 갯수를 제공하는 API가 따로 있는 것이 아니라 일반적인 요청과 그 응답에서 totalNum을 가지고오는 것 같아 말씀드립니다

mingming-mentor reviewed (approved):
수고 많으셨습니다!

seokhyeong9171 committed:
refactor: 피드백 반영

공공데이터에서 가져오는 Json을 매핑할 클래스를 만들어 매핑된 결과를 가져오도록 변경하였습니다.

기존에는 데이터의 total num을 구하고 그에 맞춰 데이터를 가져왔지만 데이터의 갯수가 0개가 되면 데이터 가져오기를 중지하는 방식으로 바꾸었습니다.

추후 food entity의 업데이트를 위해 pk를 code로 변경해 주었습니다.

seokhyeong9171 commented:
@mingming-mentor 멘토님 피드백 주신 내용 수정했습니다.
수정한 내용 한번만 더 피드백 부탁드립니다!!

mingming-mentor reviewed (approved):
훨씬 코드가 깔끔해진듯 하네요! 수고 많으셨습니다 :)
